keras_cv/models/segmentation/deeplabv3plus.py

```diff
@@ -0,0 +1,354 @@
+            spatial information for the supplied backbone, respectively.
+        weights: weights for the complete DeepLabV3Plus model. one of `None` (random
+            initialization), a pretrained weight file path, or a reference to
+            pre-trained weights (e.g. 'imagenet/classification') (see available
+            pre-trained weights in weights.py)
+        spatial_pyramid_pooling: also known as Atrous Spatial Pyramid Pooling (ASPP).
+            Performs spatial pooling on different spatial levels in the pyramid, with
+            dilation.
+        segmentation_head: an optional `tf.keras.Layer` that predict the segmentation
+            mask based on feature from backbone and feature from decoder.
+        segmentation_head_activation: default 'softmax', the activation layer to apply after
+            the segmentation head. Should be synchronized with the backbone's final activation
+    """
+
+    def __init__(
+        self,
+        classes,
+        backbone,
+        spatial_pyramid_pooling=None,
+        segmentation_head=None,
+        segmentation_head_activation="softmax",
+        input_shape=(None, None, 3),
+        input_tensor=None,
+        layer_names=(None, None),
+        weights=None,
+        **kwargs,
+    ):
+
+        if not isinstance(backbone, tf.keras.layers.Layer):
+            raise ValueError(
+                "Backbone need to be a `tf.keras.layers.Layer`, " f"received {backbone}"
+            )
+
+        if weights and not tf.io.gfile.exists(
+            parse_weights(weights, True, "deeplabv3plus")
+        ):
+            raise ValueError(
+                "The `weights` argument should be either `None` or the path to the "
+                f"weights file to be loaded. Weights file not found at location: {weights}"
+            )
+
+        inputs = utils.parse_model_inputs(input_shape, input_tensor)
+        x = inputs
+
+        if input_shape[0] is None and input_shape[1] is None:
+            input_shape = backbone.input_shape[1:]
+
+        if input_shape[0] is None and input_shape[1] is None:
+            raise ValueError(
+                "Input shapes for both the backbone and DeepLabV3Plus are `None`."
+            )
+
+        if layer_names == (None, None):
+            if "res" in backbone.name:
+                low_level_output = backbone.get_layer("v2_stack_0_block3_out").output
+                high_level_output = backbone.get_layer(
+                    "v2_stack_3_block3_out"
+                ).output
+            else:
+                raise ValueError(
+                    "Cannot default low-level and high-level layer names with a custom backbone."
+                    f"Passed layer_names: {layer_names}"
+                )
+        else:
+            low_level_output = backbone.get_layer(layer_names[0]).output
+            high_level_output = backbone.get_layer(layer_names[1]).output
```

Comment:
I'd very much prefer not to have `layer_names` in the API. Instead we could have:
`
backbone = ResNet50(...).as_backbone(levels=[2, 4])
model = DeepLabV3Plus(backbone)
`

Reply:
Mentioned this in an earlier comment - the existing API doesn't actually work for DeepLabV3. `as_backbone()` returns the last layer at that level, with 2048 channels, which would overpower the `48` from the low-level projection during addition. The paper states that either 256 or 512 should be used, which is a layer earlier, practically.
Without the low-level projections, V3Plus collapses back into V3.

Also, this would return another middle layer that isn't used. By skipping the unnecessary layers from `as_backbone()`, the model went from 39M params down to 11M params, while retaining accuracy. In #1245, @quantumalaviya proposes an update for the backbone extraction.

I guess that we could do 2048 + 256 instead of 256 + 48, skipping the 48-filter projection of the low-level layer output, and diverge from the paper, if that retains accuracy?